include/autoppl/expr_builder.hpp

```diff
@@ -0,0 +1,164 @@
+#include <autoppl/expression/distribution/bernoulli.hpp>
+
+namespace ppl {
+
+/*
+ * The purpose for these expression builders is to
+ * add extra type-safety and ease the user API.
+ */
+
+////////////////////////////////////////////////////////
+// Distribution Expression Builder
+////////////////////////////////////////////////////////
+
+namespace details {
+
+/*
+ * Converter from arbitrary (decayed) type to valid continuous parameter type 
+ * by the following mapping:
+ * - is_var_v<T> true => VariableViewer<T>
+ * - T is same as cont_raw_param_t => Constant<T>
+ * - is_var_expr_v<T> true => T
+ * Assumes each condition is non-overlapping.
+ */
+template <class T, class = void>
+struct convert_to_cont_dist_param
+{};
+
+template <class T>
+struct convert_to_cont_dist_param<T,
+    std::enable_if_t<util::is_var_v<std::decay_t<T>> && 
+                    !std::is_same_v<std::decay_t<T>, util::cont_raw_param_t> &&
+                    !util::is_var_expr_v<std::decay_t<T>>
+                    >>
+{
+    using type = expr::VariableViewer<std::decay_t<T>>;
+};
+
+template <class T>
+struct convert_to_cont_dist_param<T, 
+    std::enable_if_t<!util::is_var_v<std::decay_t<T>> && 
+                    std::is_same_v<std::decay_t<T>, util::cont_raw_param_t> &&
+                    !util::is_var_expr_v<std::decay_t<T>>
+                    >>
+{
+    using type = expr::Constant<std::decay_t<T>>;
+};
+
+template <class T>
+struct convert_to_cont_dist_param<T, 
+    std::enable_if_t<!util::is_var_v<std::decay_t<T>> && 
+                    !std::is_same_v<std::decay_t<T>, util::cont_raw_param_t> &&
+                    util::is_var_expr_v<std::decay_t<T>>
+                    >>
+{
+    using type = T;
+};
+
+template <class T>
+using convert_to_cont_dist_param_t = 
+    typename convert_to_cont_dist_param<T>::type;
+
+} // namespace details
+
+#ifndef AUTOPPL_USE_CONCEPTS
+/*
+ * Builds a Uniform expression only when the parameters
+ * are both valid continuous distribution parameter types.
+ * See var_expr.hpp for more information.
+ */
+template <class MinType, class MaxType>
+inline constexpr auto uniform(MinType&& min_expr,
+                              MaxType&& max_expr)
+{
+    using min_t = details::convert_to_cont_dist_param_t<MinType>;
+    using max_t = details::convert_to_cont_dist_param_t<MaxType>;
+
+    min_t wrap_min_expr = std::forward<MinType>(min_expr);
+    max_t wrap_max_expr = std::forward<MaxType>(max_expr);
+
+    return expr::Uniform(wrap_min_expr, wrap_max_expr);
+}
+#else
+#endif
+
+#ifndef AUTOPPL_USE_CONCEPTS
+/*
+ * Builds a Normal expression only when the parameters
+ * are both valid continuous distribution parameter types.
+ * See var_expr.hpp for more information.
+ */
+template <class MeanType, class StddevType>
+inline constexpr auto normal(MeanType&& mean_expr,
+                             StddevType&& stddev_expr)
+{
+    using mean_t = details::convert_to_cont_dist_param_t<MeanType>;
```

Comment:
I'm a little worried about how verbose and unclear this call is when it's essential to the everything working. Is there something to prevent someone from forgetting about this and breaking MH?

Reply:
If this is not there, your model won't even compile for like the basic cases like `normal(0., 1.)` xD.